pulp_container/app/serializers.py

```diff
@@ -774,65 +774,90 @@ class OCIBuildImageSerializer(ValidateFieldsMixin, serializers.Serializer):
-                )
-            data["containerfile_artifact"] = Artifact.init_and_validate(data.pop("containerfile"))
-        elif "containerfile_artifact" in data:
-            data["containerfile_artifact"].touch()
-        else:
+        if bool(data.get("containerfile", None)) == bool(data.get("containerfile_name", None)):
+            raise serializers.ValidationError(
+                _("Exactly one of 'containerfile' or 'containerfile_name' must be specified.")
+            )
+
+        if "containerfile_name" in data and "build_context" not in data:
             raise serializers.ValidationError(
-                _("'containerfile' or 'containerfile_artifact' must " "be specified.")
+                _("A 'build_context' must be specified when 'containerfile_name' is present.")
             )
-        artifacts = {}
-        if "artifacts" in data:
-            for url, relative_path in data["artifacts"].items():
-                if os.path.isabs(relative_path):
+
+        if self._is_permission_function_request():
+            # the "has_repo_or_repo_ver_param_model_or_obj_perms" permission condition function
+            # expects a "repo" or "repository_version" arguments, so we need to pass "build_context"
+            # as "repository_version" to be able to validate the permissions
+            if data.get("build_context", None):
+                data["repository_version"] = data["build_context"]
+
+            # return early, we don't need the remaining validations to check the permissions
+            return data
+
+        if containerfile := data.get("containerfile", None):
+            temp_file = PulpTemporaryFile.objects.update_or_create(file=containerfile)[0]
+            data["containerfile_tempfile_pk"] = str(temp_file.pk)
+
+        if build_context := data.get("build_context", None):
+            data["content_artifact_pks"] = []
+            repository_version = RepositoryVersion.objects.get(pk=build_context.pk)
+            content_artifacts = ContentArtifact.objects.filter(
+                content__pulp_type="file.file", content__in=repository_version.content
+            ).order_by("-content__pulp_created")
+            for content_artifact in content_artifacts.select_related("artifact").iterator():
+                if not content_artifact.artifact:
                     raise serializers.ValidationError(
-                        _("Relative path cannot start with '/'. " "{0}").format(relative_path)
+                        "It is not possible to use File content synced with on-demand "
+                        "policy without pulling the data first."
+                    )
+                containerfile_name = data.get("containerfile_name", None)
+                if containerfile_name and content_artifact.relative_path == containerfile_name:
+                    artifact = Artifact.objects.get(pk=content_artifact.artifact.pk)
```

Comment:
We can already access the artifact via the `content_artifact.artifact` attribute.